File: Engine/Android/core/src/test/kotlin/despia/engine/RecorderConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The recorder conformance runner - executes
 * OpenSource/Conformance/recorder/{lifecycle,interruption}.json through THIS runtime
 * (parity/F14-recorder.md). The TS twin is packages/kernel/test/recorder-conformance.test.ts
 * and the Swift reference is Engine/iOS/RecorderCore.swift.
 *
 * The two things that have to be identical across renderers are the metering curve (iOS
 * reports dBFS, this platform reports a 16-bit amplitude, and without one fold the same
 * voice fills the bar on one platform and not the other) and the state machine, including
 * the interruption path where every naive recorder loses a take.
 *
 * Missing corpus = loud failure.
 */
class RecorderConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/recorder/$name")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/recorder/$name not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun corpus(name: String): Map<String, Any?> {
        val root = json(corpusFile(name).readText()).foundationValue as? Map<String, Any?>
            ?: error("$name: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "$name version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun rows(doc: Map<String, Any?>, key: String): List<Map<String, Any?>> {
        val list = doc[key] as? List<Map<String, Any?>> ?: error("no $key[]")
        assertTrue(list.isNotEmpty(), "$key must not be empty")
        return list
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun curveConstantsAgreeWithCorpus() {
        val doc = corpus("lifecycle.json")
        val curve = doc["curve"] as Map<String, Any?>
        assertEquals((curve["floorDb"] as Number).toDouble(), RecorderCore.FLOOR_DB)
        assertEquals((curve["meterIntervalMs"] as Number).toInt(), RecorderCore.METER_INTERVAL_MS)
        assertEquals(doc["states"] as List<String>, RecorderCore.STATES)
    }

    @Test
    fun decibelsFoldToThePinnedMeterLevel() {
        for (case in rows(corpus("lifecycle.json"), "meter")) {
            assertEquals((case["expect"] as Number).toDouble(),
                RecorderCore.meterLevel((case["db"] as Number).toDouble()), case["name"] as? String)
        }
    }

    @Test
    fun linearAmplitudeFoldsThroughTheSameCurve() {
        for (case in rows(corpus("lifecycle.json"), "amplitude")) {
            val amplitude = (case["amplitude"] as Number).toDouble()
            val fullScale = (case["fullScale"] as Number).toDouble()
            assertEquals((case["expectDb"] as Number).toDouble(),
                RecorderCore.amplitudeDb(amplitude, fullScale), "${case["name"]}: dB")
            assertEquals((case["expectLevel"] as Number).toDouble(),
                RecorderCore.meterFromAmplitude(amplitude, fullScale), "${case["name"]}: level")
        }
    }

    @Test
    fun formatFoldAgreesWithCorpus() {
        for (case in rows(corpus("lifecycle.json"), "formats")) {
            val expect = case["expect"] as? String
            val got = RecorderCore.foldFormat(case["input"] as? String)
            if (expect == null) assertNull(got, case["name"] as? String) else assertEquals(expect, got, case["name"] as? String)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun stateMachineAgreesWithCorpus() {
        for (case in rows(corpus("lifecycle.json"), "transitions")) {
            val expect = case["expect"] as Map<String, Any?>
            val got = RecorderCore.transition(case["from"] as String, case["event"] as String)
            assertEquals(expect["state"], got.state, "${case["name"]}: state")
            assertEquals(expect["error"] as? String, got.error, "${case["name"]}: error")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun everyInterruptionSequenceWalksThePinnedStates() {
        for (case in rows(corpus("interruption.json"), "sequences")) {
            val steps = case["steps"] as List<String>
            val states = case["expect"] as List<String>
            val emits = case["emits"] as List<String?>
            assertEquals(steps.size, states.size, "${case["name"]}: one expected state per step")
            assertEquals(steps.size, emits.size, "${case["name"]}: one expected emission per step")
            var state = "idle"
            for (i in steps.indices) {
                val got = RecorderCore.transition(state, steps[i])
                state = got.state
                assertEquals(states[i], state, "${case["name"]} step ${i + 1} (${steps[i]}): state")
                assertEquals(emits[i], got.emit, "${case["name"]} step ${i + 1} (${steps[i]}): emission")
            }
        }
    }

    @Test
    fun theTakeIsNeverLostToAnInterruption() {
        var state = RecorderCore.transition("idle", "start").state
        state = RecorderCore.transition(state, "interrupt").state
        assertEquals("interrupted", state)
        val stopped = RecorderCore.transition(state, "stop")
        assertEquals("idle", stopped.state)
        assertNull(stopped.error, "stopping an interrupted take must not be refused")
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun declaredBroadcastsAreTheOnesTheMachineEmits() {
        val declared = (corpus("interruption.json")["broadcasts"] as Map<String, Any?>)
            .keys.filterNot { it.startsWith("_") }.sorted()
        val emitted = sortedSetOf<String>()
        for (from in RecorderCore.STATES) {
            for (event in listOf("start", "pause", "resume", "stop", "cancel",
                                 "interrupt", "endInterruption", "maxDuration")) {
                RecorderCore.transition(from, event).emit?.let { emitted.add(it) }
            }
        }
        assertEquals(declared, emitted.toList())
    }
}

File: Web/packages/kernel/test/recorder-conformance.test.ts
//
//  recorder-conformance.test.ts - the SHARED recorder corpus through the TS kernel
//  (OpenSource/Conformance/recorder/{lifecycle,interruption}.json). The Kotlin twin is
//  :core RecorderConformanceTest and the Swift reference is Engine/iOS/RecorderCore.swift.
//
//  The two things that have to be identical across renderers are here and nowhere else: the
//  metering curve (iOS reports dBFS, Android reports a 16-bit amplitude, and without one
//  fold the same voice fills the bar on one platform and not the other) and the state
//  machine, including the interruption path where every naive recorder loses a take.
//
//  Missing corpus = loud failure.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  RECORDER_FLOOR_DB, RECORDER_METER_INTERVAL_MS, RECORDER_STATES,
  foldRecorderFormat, recorderMeterLevel, recorderAmplitudeDb, recorderMeterFromAmplitude,
  recorderTransition,
} from "../src/recorder-core.ts";

type Dict = { [k: string]: unknown };

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/recorder");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/recorder not found");
    dir = parent;
  }
}

function corpus(name: string): Dict {
  const doc = JSON.parse(readFileSync(join(corpusDir(), name), "utf-8")) as Dict;
  assert.equal(doc["version"], 1, `${name}: version`);
  return doc;
}

function rows<T>(doc: Dict, key: string): T[] {
  const list = doc[key];
  assert.ok(Array.isArray(list) && list.length > 0, `${key} must be a non-empty array`);
  return list as T[];
}

test("recorder lifecycle: the curve constants", () => {
  const curve = corpus("lifecycle.json")["curve"] as Dict;
  assert.equal(RECORDER_FLOOR_DB, curve["floorDb"]);
  assert.equal(RECORDER_METER_INTERVAL_MS, curve["meterIntervalMs"]);
  assert.deepEqual([...RECORDER_STATES], corpus("lifecycle.json")["states"]);
});

test("recorder lifecycle: dBFS folds to the pinned meter level", () => {
  for (const row of rows<Dict>(corpus("lifecycle.json"), "meter")) {
    assert.equal(recorderMeterLevel(row["db"] as number), row["expect"], row["name"] as string);
  }
});

test("recorder lifecycle: a linear amplitude folds through the same curve", () => {
  for (const row of rows<Dict>(corpus("lifecycle.json"), "amplitude")) {
    const db = recorderAmplitudeDb(row["amplitude"] as number, row["fullScale"] as number);
    assert.equal(db, row["expectDb"], `${row["name"]}: dB`);
    assert.equal(recorderMeterFromAmplitude(row["amplitude"] as number, row["fullScale"] as number),
      row["expectLevel"], `${row["name"]}: level`);
  }
});

test("recorder lifecycle: the format fold", () => {
  for (const row of rows<Dict>(corpus("lifecycle.json"), "formats")) {
    assert.equal(foldRecorderFormat(row["input"] as string), row["expect"], row["name"] as string);
  }
});

test("recorder lifecycle: the state machine", () => {
  for (const row of rows<Dict>(corpus("lifecycle.json"), "transitions")) {
    const expect = row["expect"] as Dict;
    const got = recorderTransition(row["from"] as string, row["event"] as string);
    assert.equal(got.state, expect["state"], `${row["name"]}: state`);
    assert.equal(got.error ?? undefined, expect["error"] ?? undefined, `${row["name"]}: error`);
  }
});

test("recorder interruption: every sequence walks the pinned states and emits", () => {
  for (const row of rows<Dict>(corpus("interruption.json"), "sequences")) {
    const steps = row["steps"] as string[];
    const states = row["expect"] as string[];
    const emits = row["emits"] as (string | null)[];
    assert.equal(steps.length, states.length, `${row["name"]}: one expected state per step`);
    assert.equal(steps.length, emits.length, `${row["name"]}: one expected emission per step`);

    let state = "idle";
    for (let i = 0; i < steps.length; i += 1) {
      const got = recorderTransition(state, steps[i]);
      state = got.state;
      assert.equal(state, states[i], `${row["name"]} step ${i + 1} (${steps[i]}): state`);
      assert.equal(got.emit ?? null, emits[i], `${row["name"]} step ${i + 1} (${steps[i]}): emission`);
    }
  }
});

test("recorder interruption: the take is never lost to an interruption", () => {
  // The property in one line of code: from any interrupted take, `stop` still reaches idle
  // through the normal path, which is what makes the partial file finalisable.
  let state = recorderTransition("idle", "start").state;
  state = recorderTransition(state, "interrupt").state;
  assert.equal(state, "interrupted");
  const stopped = recorderTransition(state, "stop");
  assert.equal(stopped.state, "idle");
  assert.equal(stopped.error, undefined, "stopping an interrupted take must not be refused");
});

test("recorder interruption: the declared broadcasts are the ones the machine emits", () => {
  const declared = Object.keys(corpus("interruption.json")["broadcasts"] as Dict).filter((k) => !k.startsWith("_"));
  const emitted = new Set<string>();
  for (const from of RECORDER_STATES) {
    for (const event of ["start", "pause", "resume", "stop", "cancel", "interrupt", "endInterruption", "maxDuration"]) {
      const emit = recorderTransition(from, event).emit;
      if (emit !== undefined) emitted.add(emit);
    }
  }
  assert.deepEqual([...emitted].sort(), declared.sort());
});
